smartsim/_core/entrypoints/telemetrymonitor.py

```diff
@@ -55,17 +63,526 @@
+        self._entity = entity
+        self._sink = sink
+        self._value: t.Any = None
+        self._enabled = True
+
+    def disable(self) -> None:
+        self._enabled = False
+
+    @property
+    def enabled(self) -> bool:
+        return self._enabled
+
+    @property
+    def entity(self) -> JobEntity:
+        return self._entity
+
+    @property
+    def owner(self) -> str:
+        """The name of the SmartSim entity the collector is attached to"""
+        return self._entity.name
+
+    @property
+    def sink(self) -> Sink:
+        """The sink where collected data is written"""
+        return self._sink
+
+    @abc.abstractmethod
+    async def prepare(self) -> None:
+        """Initialization logic for a collector"""
+
+    @abc.abstractmethod
+    async def collect(self) -> None:
+        """Execute metric collection against a producer"""
+
+    @staticmethod
+    def timestamp() -> int:
+        """Return an integer timestamp"""
+        return int(datetime.datetime.timestamp(datetime.datetime.now()))
+
+    @abc.abstractmethod
+    async def shutdown(self) -> None:
+        """Execute any cleanup of resources for the collector"""
+
+
+class TaskStatusHandler(PatternMatchingEventHandler):
+    """A file listener that will notify a set of collectors when an
+    unmanaged entity has completed"""
+
+    def __init__(
+        self,
+        collector: Collector,
+        patterns: t.Optional[t.List[str]] = None,
+        ignore_patterns: t.Optional[t.List[str]] = None,
+        ignore_directories: bool = False,
+        case_sensitive: bool = False,
+    ):
+        self._collectors = [collector]
+        super().__init__(
+            patterns, ignore_patterns, ignore_directories, case_sensitive
+        )  # type: ignore
+
+    def add(self, collector: Collector) -> None:
+        self._collectors.append(collector)
+
+    def on_modified(self, event: FileModifiedEvent) -> None:
+        """Event handler for when a file or directory is modified.
+
+        :param event: Event representing file/directory modification.
+        :type event: FileModifiedEvent
+        """
+        super().on_modified(event)  # type: ignore
+        self._notify(event.src_path)
+
+    def on_created(self, event: FileCreatedEvent) -> None:
+        """Event handler for when a file or directory is created.
+
+        :param event: Event representing file/directory creation.
+        :type event: FileCreatedEvent
+        """
+        super().on_created(event)  # type: ignore
+        self._notify(event.src_path)
+
+    def _notify(self, event_src: str) -> None:
+        """Notify the collector that the entity has stopped"""
+        logger.debug(f"Processing stop event created @ {event_src}")
+
+        working_set = [item for item in self._collectors if item.enabled]
+        for col in working_set:
+            logger.debug(f"Disabling {col.entity.name}::{type(col).__name__}")
+            col.disable()
+            col.entity.set_complete()
```

Comment:
I thought the same thing and I _want_ to support that, but I had to stop somewhere. I will review and see if I can just inject the target "event" instead of having it hard-coded.

Reply:
I've subclassed `TaskStatusHandler` w/`TaskCompleteHandler` that implements _notify using the existing code. We could add a `TaskStart/Update/Etc.` subclass for other actions the same way:

```py

class TaskCompleteHandler(TaskStatusHandler):

    def _notify(self, event_src: str) -> None:
        """Notify the collector that the entity has stopped"""
        logger.debug(f"Processing stop event created @ {event_src}")

        working_set = [item for item in self._collectors if item.enabled]
        for col in working_set:
            logger.debug(f"Disabling {col.entity.name}::{type(col).__name__}")
            col.disable()
            col.entity.set_complete()
```

Reply:
Love it!! Makes sense to me!